Move WebDriverServer classes into Browser.

jgraham opened the pull request:
The code previously had a slightly odd setup where the Browser class
for WebDriver implementations was mostly using a WebDriverServer
subclass for the actual functionality, but that class and all its
subclasses were defined in a single file rather than in a
browser-specific file. This was used to put the Executor class in
charge of running the driver binary for wdspec tests, whilst the
Browser class was replaced with a dummy implementation.

It makes a lot more sense to have the WebDriver functionality directly
in the Browser class for those cases where it's used. This localises
browser-specific code to a single file rather than spreading it over
several files. It also allows wdspec tests to work like other tests
where the Browser instance owns the driver binary rather than the
executor owning the binary. That also means we don't require
boilerplate executorfoo.py files for each foo that can run wdspec
tests.

Since processes are now always owned by the browser class which lives
on the main thread, output handling is also simplified.

jgraham committed:
Move WebDriverServer classes into Browser.

The code previously had a slightly odd setup where the Browser class
for WebDriver implementations was mostly using a WebDriverServer
subclass for the actual functionality, but that class and all its
subclasses were defined in a single file rather than in a
browser-specific file. This was used to put the Executor class in
charge of running the driver binary for wdspec tests, whilst the
Browser class was replaced with a dummy implementation.

It makes a lot more sense to have the WebDriver functionality directly
in the Browser class for those cases where it's used. This localises
browser-specific code to a single file rather than spreading it over
several files. It also allows wdspec tests to work like other tests
where the Browser instance owns the driver binary rather than the
executor owning the binary. That also means we don't require
boilerplate executorfoo.py files for each foo that can run wdspec
tests.

Since processes are now always owned by the browser class which lives
on the main thread, output handling is also simplified.

foolip commented:
@DanielRyanSmith the changes here will likely affect the chrome/chromium split somehow. I'm going to review this, but if you want to take a look that would be nice too.

foolip commented on tools/wptrunner/wptrunner/browsers/android_weblayer.py:
```diff
@@ -89,8 +89,8 @@ def __init__(self, logger, binary,
                  symbols_path=None):
         """Creates a new representation of Chrome.  The `binary` argument gives
         the browser binary to use for testing."""
-        super(WeblayerShell, self).__init__(logger,
```
Are these changes a necessary part of this change? `WeblayerShell` is still around (doesn't appear elsewhere in the diff) so maybe this is a no-op change? Or was it necessary in some place and this is for consistency?

foolip replied:
Alright, let's leave it in!

foolip commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,117 @@ def __call__(self, line):
                                    command=" ".join(self.command) if self.command else "")
+
+
+class WebDriverBrowser(Browser):
```
Given that this PR has sat for so long now, is there code in here that was copied from other code now removed, and are there other changes in the meantime that need to be reflected here too?

jgraham replied:
I looked and I didn't notice anything I missed, at least.

foolip commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,117 @@ def __call__(self, line):
+        else:
+            self.base_path = base_path
+        self.env = os.environ.copy() if env is None else env
+        self._args = webdriver_args if webdriver_args is not None else []
```
Since we're juggling two binaries here, maybe:

```suggestion
        self.webdriver_args = webdriver_args if webdriver_args is not None else []
```

foolip commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,117 @@ def __call__(self, line):
+        self.webdriver_binary = webdriver_binary
+
+        self.host = host
+        self._port = port
```
`self.port` will be read later in this function, so rather than the `port` property doing it, can we just use `get_free_port()` here? Less indirection, easier to understand.

foolip replied:
I see. Well let's leave it alone.

foolip commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,117 @@ def __call__(self, line):
+        except KeyboardInterrupt:
+            self.stop()
+
+    def create_output_handler(self, cmd):
```
Is there some way in Python to mark that a method will be overridden? If not for that, this method would be a bit pointless.

foolip commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,117 @@ def __call__(self, line):
+        self._output_handler.after_process_start(self._proc.pid)
+
+        try:
+            wait_for_service((self.host, self.port))
```
Looks like a typo even on second look but isn't! (`wait_for_service` takes two arguments but the first is a tuple.)

And this is how far I got into reviewing this file before I saw that this is moved from a deleted file, the diff of which was hidden to begin with. Paying less attention ahead...

foolip commented on tools/wptrunner/wptrunner/browsers/chrome_ios.py:
```diff
@@ -45,37 +43,13 @@ def env_options():
-        return ExecutorBrowser, {"webdriver_url": self.server.url}
+    def make_command(self):
+        return [self.binary,
+                "--port=%s" % str(self.port)] + self._args
```
```suggestion
                f"--port={self.port}"] + self._args
```

That probably allow rewrapping too.

foolip commented on tools/wptrunner/wptrunner/browsers/edge.py:
```diff
@@ -101,20 +88,9 @@ def stop(self, force=False):
-        return ExecutorBrowser, {"webdriver_url": self.server.url}
+    def make_command(self):
+        return [self.webdriver_binary,
+                "--port=%s" % str(self.port)] + self._args
```
```suggestion
                f"--port={self.port}"] + self._args
```

foolip commented on tools/wptrunner/wptrunner/browsers/ie.py:
```diff
@@ -32,8 +34,16 @@ def executor_kwargs(logger, test_type, test_environment, run_info_data,
+class InternetExplorerBrowser(WebDriverBrowser):
+    def make_command(self):
+        return [self.binary,
+                "--port=%s" % str(self.port)] + self._args
```
```suggestion
                f"--port={self.port}"] + self._args
```

foolip commented on tools/wptrunner/wptrunner/browsers/safari.py:
```diff
@@ -172,11 +172,12 @@ def _find_safari_executable(self, wd_path):
-        self.server.start(block=False)
+    def make_command(self):
+        return [self.binary,
+                "--port=%s" % str(self.port)] + self._args
```
```suggestion
                f"--port={self.port}"] + self._args
```

foolip commented on tools/wptrunner/wptrunner/browsers/webkit.py:
```diff
@@ -76,35 +74,8 @@ def run_info_extras(**kwargs):
-    def executor_browser(self):
-        return ExecutorBrowser, {"webdriver_url": self.server.url}
+    def make_command(self):
+        return [self.webdriver_binary, "--port=%s" % str(self.port)] + self._args
```
```suggestion
        return [self.webdriver_binary, f"--port={self.port}"] + self._args
```

foolip commented on tools/wptrunner/wptrunner/browsers/firefox.py:
```diff
@@ -881,10 +890,52 @@ def check_crash(self, process, test):
+                               enable_webrender,
+                               chaos_mode_flags)
+        self.env["RUST_BACKTRACE"] = "1"
+        # This doesn't work with wdspec tests
```
This comment isn't removed from anywhere in this PR, but it is already in `tools/wptrunner/wptrunner/browsers/firefox.py` on `master`. Will it be in two places after this? (Some mishaps due to the PR sitting around seem likely, so I'm doing some spot checking.)

foolip commented on tools/wptrunner/wptrunner/browsers/firefox.py:
```diff
@@ -894,24 +945,13 @@ def settings(self, test):
+
+    def executor_browser(self):
+        cls, args = super().executor_browser()
+        args["marionette_port"] = self.marionette_port
```
Doesn't `self.marionette_port` need to be cast to `str`?

foolip commented on tools/wptrunner/wptrunner/browsers/servo.py:
```diff
@@ -86,3 +87,30 @@ def executor_browser(self):
+    def __init__(self, logger, binary="servo", webdriver_args=None,
+                 binary_args=None, host="127.0.0.1", env=None, port=None):
+
+        env = env if env is not None else os.environ.copy()
```
Matching what appears earlier in this diff and is easier to read:

```suggestion
        env = os.environ.copy() if env is None else env
```

foolip commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -294,3 +295,27 @@ def test_servers(self):
+    """Waits until network service given as a tuple of (host, port) becomes
+    available or the `timeout` duration is reached, at which point
+    ``socket.error`` is raised."""
+    print("Trying to connect to %s" % (addr,))
```
This `print` is added in this PR. It will print something like "Trying to connect to ('localhost', 8080)" because `addr` is a tuple.

I'm pretty sure this should use `logger` to not log unconditionally, and with that maybe fix the format with `f"Trying to connect to {addr[0]}:{addr[1]}"`

foolip commented on tools/wptrunner/wptrunner/environment.py:
```diff
@@ -294,3 +295,27 @@ def test_servers(self):
+            if e.errno != errno.ECONNREFUSED:
+                raise
+        else:
+            print("Connected to %s" % (addr,))
```
Replace `print()`?

foolip commented on tools/wptrunner/wptrunner/executors/protocol.py:
```diff
@@ -605,3 +606,56 @@ def load_reftest_analyzer(self, test, result):
+        pass
+
+
+class WdspecProtocol(ConnectionlessProtocol):
```
Isn't it weird to have `WdspecProtocol` inherit from `ConnectionlessProtocol`? It used to be `Protocol`.

Sure, subclasses can add things, but putting "connectionless" in the name does suggest that not having a connection is a core aspect of this and its subclasses.

Would it work to just implement `connect` and `after_connect` in `WdspecProtocol` instead?

foolip replied:
I see, I think the `ConnectionlessProtocol` name has confused me. Is the "connection" a direct connection between wptrunner and the browser, and not just any connection? I found this strange because WebDriver obviously involves a HTTP connection.

foolip reviewed (commented):
I like this simplification! It's very hard to have confidence that nothing was missed, so running this through all of our CI configs would be great.

DanielRyanSmith commented:
I was just doing some changes for the Chrome/Chromium split and looking at the browser classes in `tools/wptrunner/wptrunner/browsers` and thinking, "Edge and Opera have the exact same Browser class implementation as Chrome, and now I'm adding Chromium. I wonder if these can be merged..."
Glad to see I was not off base 🙂 
These changes will definitely affect my split issue. Perhaps I'll wait to change the `wptrunner` section of the implementation until this PR is merged.

jgraham committed:
Clean up handling of Firefox profiles for wdspec tests

Previously when we were using the NullBrowser and the executor process
owned the geckodriver instance, we had to create a profile and pass it
into the executor, and this was all sorts of complexity.

Now that wdspec tests work like normal tests and the Browser instance
owns the driver process, we can do a more straightforward thing and
have the FirefoxWdspecBrowser class create the profile. Then the only
thing that's needed in the executor is to add the profile path to the
capabilities. It gets this from the ExecutorBrowser instance it's
passed.

Since we always have one Browser instance per process this also means
we can always use on-disk profiles and remove the codepath that
handled creating profile zips, without risking different processes
trying to write to the same profile.

DanielRyanSmith commented on tools/wptrunner/wptrunner/browsers/webkit.py:
```diff
@@ -76,35 +74,8 @@ def run_info_extras(**kwargs):
-    def executor_browser(self):
-        return ExecutorBrowser, {"webdriver_url": self.server.url}
+    def make_command(self):
+        return [self.webdriver_binary, f"--port={self.port}]" + self._webdriver_args
```
It looks like the bracket closing the list is inside the f-string

jgraham committed:
Ensure we wait for WebDriver session to end

whimboo commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,123 @@ def __call__(self, line):
+    default_base_path = "/"
+
+    def __init__(self, logger, binary, webdriver_binary, webdriver_args=None,
+                 host="127.0.0.1", port=None, base_path="", env=None, **kwargs):
```
Shouldn't it be `base_path=None` instead?

jgraham replied:
I went with `base_path="/"` since that's the actual default value, and we aren't using the "class sets the default" behaviour anywhere (an alternative option would have been "we don't use this at all, time to remove it").

whimboo commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,123 @@ def __call__(self, line):
+        self.env = os.environ.copy() if env is None else env
+        self._webdriver_args = webdriver_args if webdriver_args is not None else []
+
+        self.url = "http://%s:%s%s" % (self.host, self.port, self.base_path)
```
Maybe just `self.url = f"http://{self.host}:{self.port}{self.base_path}"`

whimboo commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,123 @@ def __call__(self, line):
+        else:
+            self.base_path = base_path
+        self.env = os.environ.copy() if env is None else env
+        self._webdriver_args = webdriver_args if webdriver_args is not None else []
```
Is there any reason for the underscore here? Why not `self.webdriver_args`?

DanielRyanSmith replied:
The underscore in this situation is pythonic syntax for a `private` instance variable like you might see in other object-oriented languages. Since the functionality for private variables does not exist in the language, this is the way that devs flag an instance variable as private. No actual functionality changes - just for readability!

whimboo replied:
Sure, but that was not my actual point. Sorry that it was unclear formulated. We have `﻿﻿﻿self.webdriver_binary` on this class and the same style we should use for the arguments. I do not see a specific getter which would require us to use an underscore.

DanielRyanSmith replied:
I see, I misinterpreted your original comment!

whimboo commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -275,3 +280,123 @@ def __call__(self, line):
+        except Exception:
+            self.logger.error(
+                "WebDriver was not accessible "
+                "within the timeout:\n%s" % traceback.format_exc())
```
Maybe use a formatting string?

whimboo commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -1269,5 +1269,9 @@ def _render(self, protocol, url, timeout):
     def __init__(self, logger, browser, *args, **kwargs):
         super().__init__(logger, browser, *args, **kwargs)
+
+        args = self.capabilities["moz:firefoxOptions"].setdefault("args", [])
+        args.extend(["--profile", self.browser.profile])
```
Why aren't we going to use [`moz:firefoxOptions.profile`](https://developer.mozilla.org/en-US/docs/Web/WebDriver/Capabilities/firefoxOptions#profile) and pass a base64 encoded profile? Doesn't this change break our support for Android?

jgraham replied:
Geckodriver does the right thing on android, I think? https://searchfox.org/mozilla-central/source/testing/geckodriver/src/capabilities.rs#422 sets the `profile` field of `FirefoxOptions` if `--profile` is passed. https://searchfox.org/mozilla-central/source/testing/geckodriver/src/browser.rs#184 passes that path down to `handler.prepare` and https://searchfox.org/mozilla-central/source/testing/geckodriver/src/android.rs#366 pushes the profile from the local system to the device.

whimboo replied:
Discussed last week with @jgraham and that is indeed fine. We don't need the base64 encoded profile feature for wpt and as such it's possible to pass profile through the Firefox command line argument, which are then picked-up by geckodriver.

whimboo commented on tools/wptrunner/wptrunner/browsers/firefox.py:
```diff
@@ -870,6 +873,37 @@ def start(self, group_metadata, **kwargs):
+                    self.logger.debug("Connecting to /status failed")
+                    break
+                if res.status != 200:
+                    self.logger.debug(f"Connecting to /status gave status {res.status}")
```
Would it be ok to have the full URL in the status message?

whimboo commented on tools/wptrunner/wptrunner/browsers/firefox.py:
```diff
@@ -870,6 +873,37 @@ def start(self, group_metadata, **kwargs):
         super().start(group_metadata, **kwargs)
 
+    def stop(self, force=False):
+        # Initially wait for any WebDriver session to cleanly shutdown
```
Can you please explain what the benefits of these extra checks are? Maybe extend the comment to explain it in the code.

whimboo commented on webdriver/tests/support/fixtures.py:
```diff
@@ -31,6 +31,15 @@ def pytest_configure(config):
     )
 
 
+def pytest_sessionfinish(session, exitstatus):
```
Good catch!

foolip reviewed (approved):
All of my comments have been addressed. Thanks @jgraham!

jgraham committed:
Remove unused webdriver_server file

DanielRyanSmith commented:
I opened [a draft CL](https://chromium-review.googlesource.com/c/chromium/src/+/3381499) for these changes on Chromium to test if there will be any issues in the CI do and it seems like the coast is clear (if I handled it properly 😅)

jgraham committed:
Reimplement support for wdspec tests running in Firefox for android

whimboo reviewed (approved):
I think that this looks good now. Thanks @jgraham!

whimboo reviewed (approved):
I checked the recently added commit by @jgraham and it looks fine to me. It will actually enable us to finally continue enabling wdspec tests for Firefox on Android.

Due to the forced push it makes it hard to find other code changes. @jgraham, so if there is something I have to take a look at please point me to these. Otherwise we could get it merged if all checks are passing.

Thanks!

jgraham commented:
OK, the Safari issue seems to be a timeout, and looking at the logs the tests seem to be running correctly, so I'm going to go ahead and merge this.
